### PR TITLE
Enable nullability in Application

### PR DESCRIPTION
**Pull request by gpetrou**
## Proposed changes

- Enable nullability in Application.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8324)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.ThreadContext.cs [hunk 1]
```diff
@@ -1013,7 +1013,7 @@ private unsafe void RunMessageLoopInner(msoloop reason, ApplicationContext conte
 
                     ApplicationContext = context;
 
-                    ApplicationContext.ThreadExit += new EventHandler(OnAppThreadExit);
+                    ApplicationContext!.ThreadExit += new EventHandler(OnAppThreadExit);
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.ThreadContext.cs:1016]: Looks like this method need to be cleaned but can go in a separate PR.

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs [hunk 2]
```diff
@@ -230,11 +228,11 @@ public static string CompanyName
                         // won't work with MC++ see GetAppMainType.
                         if (s_companyName is null || s_companyName.Length == 0)
                         {
-                            Type t = GetAppMainType();
+                            Type? t = GetAppMainType();
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.cs:231]: nit:  can you change `initials` to full name  t->type  wherever we touch the source?

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.cs [hunk 3]
```diff
@@ -839,7 +840,7 @@ internal static void EndModalMessageLoop()
         /// <summary>
         ///  Overload of <see cref="Exit(CancelEventArgs)"/> that does not care about e.Cancel.
         /// </summary>
-        public static void Exit() => Exit(null);
+        public static void Exit() => Exit(null!);
```
Comment (dreddy-work) [inline src/System.Windows.Forms/src/System/Windows/Forms/Application.cs:843]: Any reason to use `!` ?  instead of changing the method and it was already expecting `null` events.

File: src/System.Windows.Forms/src/System/Windows/Forms/Application.ThreadContext.cs (hunk 1 above)
Comment (gpetrou) [reply src/System.Windows.Forms/src/System/Windows/Forms/Application.ThreadContext.cs:1016]: It is a bit confusing, but when we have `reason` equal to `msoloop.Main` this should not be null.

**Review by dreddy-work — APPROVED** [review]
